### PR TITLE
JS: improve join-order for HTTP::isDecoratedCall

### PR DESCRIPTION
**Pull request by erik-krogh**
My [recent PR](https://github.com/github/codeql/pull/4275) apparently got a bad join-order in some of the minor edits made towards the end. 

This PR fixes that join-order (and uses `DataFlow::FunctionNode` instead of `Function`).

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/HTTP.qll [hunk 1]
```diff
@@ -240,14 +240,22 @@ module HTTP {
...
   }
 
+  /**
+   * Gets a function that returns the route-handler-candidate `routeHandler`.
+   */
+  pragma[noinline]
+  private DataFlow::FunctionNode returnsARouteHandler(HTTP::RouteHandlerCandidate routeHandler) {
+    routeHandler = result.getAReturn().getALocalSource()
+  }
+
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/HTTP.qll:258]: I found the naming, documentation and implementation to be a bit mismatched. Can we use something like this canonical form instead?

```suggestion
  /**
   * Holds if `fun` returns the route-handler-candidate `routeHandler`.
   */
  pragma[noinline]
  private predicate returnsRouteHandler(DataFlow::FunctionNode fun, HTTP::RouteHandlerCandidate routeHandler) {
    routeHandler = result.getAReturn().getALocalSource()
  }

```

File: javascript/ql/src/semmle/javascript/frameworks/HTTP.qll [hunk 2]
```diff
@@ -240,14 +240,22 @@ module HTTP {
...
   private predicate isDecoratedCall(DataFlow::CallNode call, DataFlow::FunctionNode decoratee) {
     // indirect route-handler `result` is given to function `outer`, which returns function `inner` which calls the function `pred`.
-    exists(int i, Function outer, Function inner |
+    exists(int i, DataFlow::FunctionNode outer, HTTP::RouteHandlerCandidate inner |
       decoratee = call.getArgument(i).getALocalSource() and
-      outer = call.getACallee() and
-      inner = outer.getAReturnedExpr() and
-      isAForwardingRouteHandlerCall(DataFlow::parameterNode(outer.getParameter(i)), inner.flow())
+      outer.getFunction() = call.getACallee() and
+      outer = returnsARouteHandler(inner) and
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/HTTP.qll:246]: (see below)
```suggestion
      returnsRouteHandler(outer, inner) and
```

File: javascript/ql/src/semmle/javascript/frameworks/HTTP.qll [hunk 3]
```diff
@@ -240,14 +240,22 @@ module HTTP {
    */
   private predicate isDecoratedCall(DataFlow::CallNode call, DataFlow::FunctionNode decoratee) {
     // indirect route-handler `result` is given to function `outer`, which returns function `inner` which calls the function `pred`.
-    exists(int i, Function outer, Function inner |
+    exists(int i, DataFlow::FunctionNode outer, HTTP::RouteHandlerCandidate inner |
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/frameworks/HTTP.qll:243]: I should have called this out earlier, but can we perhaps include a source code example that relates `decoratee`, `outer` and `inner`?

**Review by esbena — COMMENTED** [review]
I have some clarification requests.

**Comment by erik-krogh** [comment]
[An evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1600736835477) came back less interesting than I had expected (given that `isDecoratedCall` on `react` went from 16s to negligible on my laptop).   
The first column is before my [earlier PR](https://github.com/github/codeql/pull/4275), the second column is the bad join order, the third column is the fixed join-order.   
(You can ignore the reruns).

**Comment by esbena** [comment]
I suspect you did a different experiment on your laptop.

The 16 seconds have turned into 200 seconds on a different machine, that seems a bit much for identical runs.

> react went from 16s to negligible

> `react            200                     201                     1.02791  1.02186                 1.005     181                     1.00339  1.00527  0.905`

**Comment by erik-krogh** [comment]
> I suspect you did a different experiment on your laptop.

Yes. I was talking about the "Clause timing report", where the `isDecoratedCall` predicate went from 16s to not appearing in the report.

**Comment by esbena** [comment]
I see.
So we have one very local measurement that shows that this PR does has a large effect, and then a larger experiment that shows that the cost basically just moves elsewhere. I am sorry, but I think that is a bit thin for adding a new `pragma`.

**Comment by esbena** [comment]
NVM.
I missed the third column (again). There is definitely an effect. Lets merge.

**Review by esbena — APPROVED** [review]

**Pull request merged**